rosidl_generator_type_hash/rosidl_generator_type_hash/__init__.py

```diff
@@ -0,0 +1,314 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import hashlib
+import json
+from pathlib import Path
+import sys
+from typing import List
+
+from rosidl_parser import definition
+from rosidl_parser.parser import parse_idl_file
+
+
+def generate_type_hash(generator_arguments_file: str) -> List[str]:
+    with open(generator_arguments_file, 'r') as f:
+        args = json.load(f)
+    package_name = args['package_name']
+    output_dir = Path(args['output_dir'])
+    idl_tuples = args['idl_tuples']
+    include_paths = args.get('include_paths', [])
+
+    # Lookup for directory containing pregenerated .json.in files
+    include_map = {
+        package_name: output_dir
+    }
+    for include_tuple in include_paths:
+        include_parts = include_tuple.rsplit(':', 1)
+        assert len(include_parts) == 2
+        include_package_name, include_base_path = include_parts
+        include_map[include_package_name] = Path(include_base_path)
+
+    generated_files = []
+    hashers = []
+
+    # First generate all .json.in files (so referenced types can be used in expansion)
+    for idl_tuple in idl_tuples:
+        idl_parts = idl_tuple.rsplit(':', 1)
+        assert len(idl_parts) == 2
+        locator = definition.IdlLocator(*idl_parts)
+        try:
+            idl_file = parse_idl_file(locator)
+        except Exception as e:
+            print('Error processing idl file: ' +
+                  str(locator.get_absolute_path()), file=sys.stderr)
+            raise(e)
+
+        idl_rel_path = Path(idl_parts[1])
+        generate_to_dir = (output_dir / idl_rel_path).parent
+        generate_to_dir.mkdir(parents=True, exist_ok=True)
+
+        hasher = InterfaceHasher.from_idl(idl_file)
+        generated_files.extend(
+            hasher.write_json_in(output_dir))
+        hashers.append(hasher)
+
+    # Expand .json.in and generate .sha256 hash files
+    for hasher in hashers:
+        generated_files.extend(
+            hasher.write_json_out(output_dir, include_map))
+        generated_files.extend(
+            hasher.write_hash(output_dir))
+
+    return generated_files
+
+
+# This mapping must match the constants defined in type_description_interfaces/msgs/FieldType.msg
+# TODO(emersonknapp)
+# There is no FieldType.msg definition for the following rosidl_parser.definition types
+# * SIGNED_NONEXPLICIT_INTEGER_TYPES = short, long, long long
+# * UNSIGNED_NONEXPLICIT_INTEGER_TYPES = unsigned short, unsigned long, unsigned long long
+FIELD_TYPES = {
```

Comment:
Actually, for the purposes of hash generation, would it be better to just put the type_ids as strings instead of their corresponding indices? That way it'd be robust to any changes in the type IDs (and we can always defer parsing down the line if we ever want to use the jsons otherwise?)

I think the only tradeoff aside from deferred parsing will just be a couple extra characters in the json representation

Reply:
Remembering that this implementation needs to match json output with https://github.com/ros2/rcl/pull/1027/files#diff-10336fa41b3ee4ac638d837fe09e40ee541e5efe2362885769fb8e783dad1047R113 - would you suggest that the `rcl` implementation use a stringifying operation for the type_id?

Reply:
I think I'd like to push back on this. If the ID constants in `type_description_interfaces/msg/FieldType.msg` change, then I think the RIHS hashing version must also be incremented, this is a breaking change to the hashing protocol.

Reply:
Hmm, ok! No problems then